### ISSUE TITLE
ml_save fails in yarn-client mode

### ISSUE THREAD

**Issue by krisgeus** [rstudio/sparklyr#336]
`ml_save(mymodel, "aDirectoryInMyHDFSUserDir")`

expected a directory called "aDirectoryInMyHDFSUserDir" in /users/<username> on HDFS with the model saved.

got: 
`Error in gzfile(file, mode) : cannot open the connection
In addition: Warning message:
In gzfile(file, mode) :
  cannot open compressed file 'aDirectoryInMyHDFSUserDir/metadata.rds', probable reason 'No such file or directory'`

Problem is that the `aDirectoryInMyHDFSUserDir` directory is created in HDFS but the saveRDS function looks for it on the machine where rstudio server is running probably in the current working directory.

**Comment by krisgeus** [comment]
Also tried in local[*] mode but writing to HDFS is the default in my spark installation also for local mode.
Specifying the file as "file:///<localdir>" also didn't work

**Comment by verty-1** [comment]
I have this exact same issue for ml_save in sparklyr, running Spark 2.1.0.2.6.0.3 in yarn-client mode.  The model is class "ml_model_random_forest" .   It appears only portions of the model are written to hdfs.  Same issue with and without full path names.  

> ml_save(ml16_rf, file = "/user/qa/keith/at5")
Error in gzfile(file, mode) : cannot open the connection
In addition: Warning message:
In gzfile(file, mode) :
  cannot open compressed file '/user/qa/keith/at5/metadata.rds', probable reason 'No such file or directory'

**Comment by verty-1** [comment]
Here is a viable work-around for those not "in the know".   The issue occurs for both local and yarn-client mode.  View the code for ml_save in R (sparklyr version 0.5.4).  The ml_save command writes spark data to HDFS, but the saveRDS data is written to the R session directory.  The saveRDS fails because the subdirectory created on HDFS does not exist in the R session directory.   To work around this, create a new function "mlsave" like this below, being aware the model files are being written in two places and pathname awareness is important.  Then, ml_load works properly without modification.

mlsave <- function (model, file) {
    ensure_scalar_character(file)
    file <- path.expand(file)
    invoke(model$.model, "save", file)
    r <- model
    r$.model <- NULL
    dir.create(file)  ## Added this line of code to ml_save
    saveRDS(r, file = file.path(file, "metadata.rds"))
    file
        }

**Comment by rexhu100** [comment]
@verty-1 kudos for looking into the source the function, but I have tried your implementation, and it doesn't work either. 

I did a little digging, and I found that the problem is that `saveRDS` function cannot connect to HDFS, and the `save` method in `invoke` cannot write to the local file system. My solution to this is to save the `metadata.rds` file to the `/tmp` folder first, and then upload it to the HDFS. My source is given below

mlsave <- function (model, file) {
  ensure_scalar_character(file)
  file <- path.expand(file)
  invoke(model$.model, "save", file)
  r <- model
  r$.model <- NULL
  \# Here is what I changed-------------------
  saveRDS(r, "/tmp/metadata.rds")
  system(
    paste("hadoop fs -put /tmp/metadata.rds", file)
  )
  \#-------------------------------------------
  file
}
